wp_api/src/reqwest_request_executor.rs

```diff
@@ -143,3 +159,227 @@ impl RequestExecutor for ReqwestRequestExecutor {
+    fn from(error: &ResolveError) -> Self {
+        // Future improvement: We could probably detect when the domain is valid, but
+        // there's no DNS record for the provided hostname
+
+        RequestExecutionErrorReason::NonExistentSiteError {
+            error_message: Some(error.to_string()),
+            suggested_action: None,
+        }
+    }
+}
+
+/// Converts from the Hyper frameworks underlying errors to a RequestExecutionErrorReason
+impl From<&HyperError> for RequestExecutionErrorReason {
+    fn from(error: &HyperError) -> Self {
+        if let Some(http2_error) = error.find::<Http2Error>() {
+            // TODO: We can probably handle more cases here, such as:
+            // - Connection reset
+
+            return RequestExecutionErrorReason::GenericError {
+                error_message: http2_error.to_string(),
+            };
+        }
+
+        if error.is_closed() {
+            return RequestExecutionErrorReason::GenericError {
+                error_message: error.to_string(),
+            };
+        }
+
+        if error.is_incomplete_message() {
+            // The server terminated the connection unexpectedly
+            return RequestExecutionErrorReason::GenericError {
+                error_message: error.to_string(),
+            };
+        }
+
+        RequestExecutionErrorReason::GenericError {
+            error_message: error.to_string(),
+        }
+    }
+}
+
+trait ExaminableError {
+    fn as_io_error(&self) -> Option<&std::io::Error>;
+    fn as_dns_error(&self) -> Option<&ResolveError>;
+    fn as_tls_error(&self) -> Option<&TlsError>;
+    fn as_hyper_error(&self) -> Option<&HyperError>;
+}
+
+impl ExaminableError for reqwest::Error {
+    fn as_io_error(&self) -> Option<&std::io::Error> {
+        self.find::<std::io::Error>()
+    }
+
+    fn as_dns_error(&self) -> Option<&ResolveError> {
+        self.find::<ResolveError>()
+    }
+
+    fn as_tls_error(&self) -> Option<&TlsError> {
+        if let Some(error) = self.as_io_error() {
+            let Some(inner_error) = error.get_ref() else {
+                println!("No inner error found for {:?}", self.url());
+                return None;
+            };
+
+            let Some(io_error) = inner_error.downcast_ref::<std::io::Error>()?.get_ref() else {
+                println!("No inner error found for {:?}", self.url());
+                return None;
+            };
+
+            let Some(tls_error) = io_error.downcast_ref::<TlsError>() else {
+                println!("No inner error found for {:?}", self.url());
+                return None;
+            };
+
+            return Some(tls_error);
+        }
+
+        None
+    }
+
+    fn as_hyper_error(&self) -> Option<&HyperError> {
+        self.find::<HyperError>()
+    }
+}
+
+// It's probably possible to have a single implementation for all of these, but we can do that later
+trait FindsError {
+    fn find<E: Error + 'static>(&self) -> Option<&E>;
```

Comment:
Removed the `'a` lifetime from the original implementation because `clippy` was complaining about it.